### DISCUSSION TITLE
Implementing a new variable that changes dynamically in simulation

### DISCUSSION THREAD

**Discussion by TimThacker** [CURENT/andes#355] [Modeling] [answered]
I am attempting to implement a coordinating signal to a governor model in ANDES, where the coordinating signal changes dynamically during TDS. 

Specifically, I am attempting to modify the governor speed reference in the IEESGO governor model from omega_ref to (omega_ref + uc) where uc is the coordinating signal.

I believe the implementation would require the creation of a new variable and some way to define that variable. I am struggling with both the variable implementation and understanding how to properly update this variable in TDS. Ideally the signal would be proportional to the power output of a renewable energy source (DFIG wind turbine, etc) however I would be happy with a simple proof of concept where at some time it changes value discretely.

Where should I begin?

**Comment by cuihantao** [comment, marked as answer]
Hi Tim,

Changing the speed reference of a turbine governor model depends on the specific model. For IEEGSO, looking at https://www.powerworld.com/WebHelp/Content/TransientModels_PDF/Generator/Governor/Governor%20IEESGO.pdf, we can tell that it provides the `paux` signal as an auxiliary input. 

In the ANDES implementation of IEESGO, `paux` is inherited from the `TGBase` class. The definition of `paux` is found here:

https://github.com/cuihantao/andes/blob/901a4fb7cda336ab3adec2cd47b63902e0ce803c/andes/models/governor/tgbase.py#L108-L112

To update `paux`, follow this example at https://docs.andes.app/en/latest/_examples/ex8.html. Whenever you update `paux0`, make sure write to its `v` field in place. See https://docs.andes.app/en/latest/_examples/ex2.html. 

The IEESGO model does not provide a way to update the speed reference, but I can point you to a feasible solution if `paux` is not what you are looking for.

**Reply by TimThacker** [reply]
Thank you for the quick response. Unfortunately, paux is not what I am looking for.

I am attempting to add a new variable to the first differential equation representing the IEESGO model. The addition of the coordinating signal _uc_ can be seen in equation 6 here: https://arxiv.org/ftp/arxiv/papers/2009/2009.07890.pdf
In this application, a reduced version of the IEESGO governor is used, where K2=K3=0. I have already created and tested this reduced model in a fork of ANDES.

You note that changing paux is a way one can input from external programs such as reinforcement learning. I am attempting a similar task, but through a modification of the model - the modification representing a coordinating signal.

To properly implement this, would I begin by adding _uc_ as an attribute in the TGBase class? 

Thank you for your help, I have been working with MATLAB recently and I am not as familiar with Python, however Python has more opportunities for the application of reinforcement learning to power systems.

**Reply by cuihantao** [reply]
In the case where you need to modify the speed reference, change the line here:

https://github.com/cuihantao/andes/blob/901a4fb7cda336ab3adec2cd47b63902e0ce803c/andes/models/governor/ieesgo.py#L70

to `u='ue * (uomega0 + omega - wref)',`, where `uomega0` is a `ConstService` initialized to 0. `uomega0` is basically the `uc`, but I named it in a way to denote a `ConstService` that modifies omega.

`uomega0` will need to be updated during your simulation loops.

**Reply by cuihantao** [reply]
You might add `domega0` to TGBase, but in the future, turbine governor model will need to be updated individually to include this constant. I haven't seen a lot of cases where the speed reference is updated; more often, the power reference is updated.
